Publish level 2+3 patterns as a gitbook automatically (#242)
gruetter/innerSourceTest: book/introduction.md

# Introduction

<img src="./innersource-patterns-book-cover.jpg" title="InnerSource Patterns">

{% hint style="info" %}
This is an early release of the InnerSource Patterns book.

> You may still find broken link, spelling mistakes, or other errors in this book.

Reviewer: ```suggestion
You may still find broken links, spelling mistakes, or other errors in this book.
```

> Please help us to fix them to produce the best book possible :) Learn here how to [contribute to this book](../book/contribute-to-this-book.md).

Reviewer: ```suggestion
Please help us to fix them to produce the best book possible :). Learn here how to [contribute to this book](../book/contribute-to-this-book.md).
```

{% endhint %}

Welcome to the **InnerSource Patterns book**.

> This book contains InnerSource best practices codified in a specific format to make it easy to understand, evaluate, and reuse them. We call this reuse format a **pattern**.

Reviewer: Nitpicky question: do we _reuse_ or rather _apply_ patterns? I think the pattern can be _applied_ and the solution described in it is _reused_.

Author: Good point! Changing it to "apply".

> This book contains InnerSource best practices codified in a specific format to make it easy to understand, evaluate, and reuse them. We call this reuse format a **pattern**.

Author: ```suggestion
This book contains InnerSource best practices codified in a specific format to make it easy to understand, evaluate, and apply them in your context. We call this format a **pattern**.
```

> These patterns have been collected by the [InnerSource Commons](http://innersourcecommons.org) community over many years. The most mature of these patterns have been published in this book.

Reviewer: By _mature_, are we referring to _proven_? If so, should we mention this to make sure people understand that these are _tried and proven_ and not just some ideas?

Author: We have changed the pattern maturity levels in the meantime. We now have 3 levels ([full details](https://github.com/InnerSourceCommons/InnerSourcePatterns/blob/master/meta/contributor-handbook.md#in-a-nutshell)):
- 1 initial
- 2 structured
- 3 validated

Patterns of level 2 or 3 are published in the book. So far we don't have any patterns in level 3 yet though :)

> These patterns have been collected by the [InnerSource Commons](http://innersourcecommons.org) community over many years. The most mature of these patterns have been published in this book.

Author: ```suggestion
These patterns have been collected by the [InnerSource Commons](http://innersourcecommons.org) community over many years. The most mature patterns have been published in this book. Mature in this context means that each pattern has been reviewed by members of the community, and has at least one known instance where this pattern has been used.
```

I think that pointing to the Contributor Handbook here to explain the levels and which ones are in the book would be too much for new readers.

Therefore I modified the description to explain briefly what we mean with _mature_ in this context.

Would that work?

In this introduction we explain [what InnerSource is](#what-is-innersource), [what a pattern is](#what-are-innersource-patterns), and [how to use these patterns](#how-can-you-use-innersource-patterns) in your organization.

If you are using InnerSource in your company already and want to contribute your experiences to this book we would love to [welcome your contributions](../book/contribute-to-this-book.md)!

## What is InnerSource?

We define InnerSource as:

> > The use of open source principles and practices for software development within the confines of a company.

Reviewer: ```suggestion
> The use of open source principles and practices for software development within the confines of an organisation.
```

This is how we define it on innersourcecommons.org.

> InnerSource takes the lessons learned from developing open source software and applies them to the way companies develop software internally. As developers have become accustomed to working on world class open source software, there is a strong desire to bring those practices back inside the firewall and apply them to software that companies may be reluctant to release.

Reviewer: > bring those practices back inside the firewall

Bringing them back implies that they were there previously. I'd argue that's seldomly if ever the case. Also, I'm not familiar with the term "inside the firewall". 

Suggestion: "... there is a strong desire to apply those practices inside their organisations to software that they may be reluctant to or can't release as Open Source."

Author: I got this text from https://innersourcecommons.org, section "What is InnerSource?" :)

I am happy to change it, but maybe we should change innersourcecommons.org as well then?

> For companies building mostly closed source software, InnerSource can be a great tool to help break down silos, encourage internal collaboration, accelerate new engineer on-boarding, and identify opportunities to contribute software back to the open source world.

Reviewer: In my experience (thus far), silos never quite go away - they are just made more "permeable". Not sure we should mention this here but I feel we need to be careful setting the right expectations.

Author: Same as above: I got this text from https://innersourcecommons.org, section "What is InnerSource?"

Author: After reviewing all other comments, I noticed that this comment and the two comments below are all modifying text that I copied from somewhere else within the ISC.

In order to get the book released quickly I would therefore propose the following:
- we resolve these 3 comments (i.e. leave the text as is)
- get the book released

After that if you want to improve these descriptions, maybe you could open new PRs against the innersourcecommons.org or InnerSourcePatterns repo, and then we can port those changes into the book after that gets merged.

What do you think?

Author: @gruetter if you can give me the green light on the approach highlighted in my last comment above, then I could get the first version of the book released over the weekend. So if you have any time to review this, would be much appreciated.

## What are InnerSource Patterns?

> Patterns are a way of describing a repeatable, proven solution to a problem with a context. They follow a simple form that helps people wanting to implement the solution to understand the constraints on the problem, the forces that must be balanced and the resulting context (the situation you are left with after the solution is applied).

Reviewer: ```suggestion
Patterns are a way of describing a repeatable, proven solution to a problem within a context. They follow a simple form that helps people wanting to implement the solution to understand the constraints on the problem, the forces that must be balanced and the resulting context - the situation created by applying the solution.
```

> In inner sourcing, patterns can provide a way for the InnerSource Commons participants to concisely share information with each other, improving the practice of inner sourcing. Each of the patterns are divided into Title, Problem Statement, Context, Forces, and Solutions as their main sections.

Reviewer: ```suggestion
In inner sourcing, patterns can provide a way for the InnerSource Commons participants to concisely share information with each other, improving the practice of inner sourcing. Pattern descriptions are subdivided into the following sections: Title, Patlet, Problem Statement, Context, Forces, Solutions and Resulting Context.
```

Author: This text comes from here: https://github.com/InnerSourceCommons/InnerSourcePatterns#what-are-innersource-patterns

But again, I am happy to change it :)

* [`What are patterns?` Youtube videos](http://bit.ly/innersource_patterns_videos) - Watch a set of 2-5 min youtube videos explaining InnerSource Patterns
* [Pattern Discussion Webinar](https://youtu.be/i-0IVhfRVFU) - We held a webinar 2017-03-16 to live-discuss a donut pattern (go to 24:30 for the discussion). This is an illustration of the review process we follow. Also see the [June 1, 2017 O'Reilly Webinar on InnerSource Patterns](http://www.oreilly.com/pub/e/3884).
* [Pattern Template](../meta/pattern-template.md) - View a skeleton inner source pattern to get an idea on what goes into a new pattern!
* [Introduction to InnerSource Patterns (2016 Fall Summit presentation)](https://drive.google.com/open?id=0B7_9iQb93uBQbnlkdHNuUGhpTXc) - *Tim Yao and Padma Sudarsan* (PDF). Detailed pattern background and examples -- Get a detailed understanding of why and how to interact with our patterns. Also see the [Introduction to InnerSource Patterns (2017 Fall Summit)](https://drive.google.com/open?id=0B7_9iQb93uBQWmYwMFpyaGh4OFU) *Tim Yao and Bob Hanmer* (PDF).

## How can you use InnerSource Patterns?

> Patterns must be used in a thoughtful manner. They cannot be blindly applied. In most cases, you will need to adapt the given solution to your own situation; but the information given in the pattern, defining the context (immovable constraints) and forces (constraints that can be changed and balanced against each other), should help you do this. Note that you will also need to determine if there are additional constraints (company context and company forces) that apply to your particular company/organization that must be added to the pattern (as a kind of filter). These additional constraints may require additional solution steps to be applied.

Reviewer: Proposal: s/immovable/invariable or s/immovable/fixed or s/immovable/immutable

Reviewer: > Note that you will also need to determine if there are additional constraints (company context and company forces) that apply to your particular company/organization that must be added to the pattern

Not sure I grok this. How would a potential user _add_ a constraint to the pattern? Do we rather mean: "... that must be considered before applying the pattern"?

Author: This description comes from https://github.com/InnerSourceCommons/InnerSourcePatterns#how-can-you-use-innersource-patterns

> The pattern form is useful for describing proven patterns but it can also be used for *brainstorming solutions* where patterns are not yet established, since the form gives a structured way for thinking about a problem. You could also create a *donut pattern* (filling in the problem, context, forces and resulting context fields but leaving the solution blank) as a way of asking the InnerSource Commons community for help (to find a proven solution or to brainstorm things to try).

Reviewer: ```suggestion
The pattern form is useful for describing proven solutions but it can also be used for *brainstorming new solutions* where patterns are not yet established. This is, because the anatomy of a pattern provides a frame for thinking about a problem in a structured manner. You could also create a *donut pattern* (filling in the problem, context, forces and resulting context fields but leaving the solution blank) as a way of asking the InnerSource Commons community for help (to find a proven solution or to brainstorm things to try).
```

Author: Great change! Merging.
